File: Signal/src/ViewControllers/ConversationView/ConversationScrollButton.m
//
//  Copyright (c) 2020 Open Whisper Systems. All rights reserved.
//

#import "ConversationScrollButton.h"
#import "UIFont+OWS.h"
#import "UIView+OWS.h"
#import <SignalMessaging/SignalMessaging-Swift.h>
#import <SignalMessaging/Theme.h>

NS_ASSUME_NONNULL_BEGIN

@interface ConversationScrollButton ()

@property (nonatomic) NSString *iconName;
@property (nonatomic) UIImageView *iconView;
@property (nonatomic) UIView *circleView;
@property (nonatomic) UIView *shadowView;

@property (nonatomic) UIView *unreadBadge;
@property (nonatomic) UILabel *unreadLabel;

@end

#pragma mark -

@implementation ConversationScrollButton

- (nullable instancetype)initWithIconName:(NSString *)iconName
{
    self = [super initWithFrame:CGRectZero];
    if (!self) {
        return self;
    }

    self.iconName = iconName;

    [self createContents];

    [NSNotificationCenter.defaultCenter addObserver:self
                                           selector:@selector(themeDidChange:)
                                               name:ThemeDidChangeNotification
                                             object:nil];

    return self;
}

+ (CGFloat)circleSize
{
    return ScaleFromIPhone5To7Plus(35.f, 40.f);
}

+ (CGFloat)buttonSize
{
    return self.circleSize + 2 * 15.f;
}

- (void)themeDidChange:(NSNotification *)notification
{
    [self updateColors];
}

- (void)createContents
{
    UIImageView *iconView = [UIImageView new];
    self.iconView = iconView;
    iconView.userInteractionEnabled = NO;

    const CGFloat circleSize = self.class.circleSize;

    UIView *shadowView = [[OWSCircleView alloc] initWithDiameter:circleSize];
    self.shadowView = shadowView;
    shadowView.userInteractionEnabled = NO;
    shadowView.layer.shadowOffset = CGSizeMake(0, 0);
    shadowView.layer.shadowRadius = 4;
    shadowView.layer.shadowOpacity = 0.05f;
    shadowView.layer.shadowColor = UIColor.blackColor.CGColor;

    UIView *circleView = [[OWSCircleView alloc] initWithDiameter:circleSize];
    self.circleView = circleView;
    circleView.userInteractionEnabled = NO;
    circleView.layer.shadowOffset = CGSizeMake(0, 4.f);
    circleView.layer.shadowRadius = 12.f;
    circleView.layer.shadowOpacity = 0.3f;
    circleView.layer.shadowColor = UIColor.blackColor.CGColor;

    UIView *unreadBadge = [UIView new];
    self.unreadBadge = unreadBadge;
    unreadBadge.userInteractionEnabled = NO;
    unreadBadge.layer.cornerRadius = 8;
    unreadBadge.clipsToBounds = YES;

    UILabel *unreadCountLabel = [UILabel new];
    self.unreadLabel = unreadCountLabel;
    unreadCountLabel.font = [UIFont systemFontOfSize:12];
    unreadCountLabel.textColor = UIColor.ows_whiteColor;
    unreadCountLabel.textAlignment = NSTextAlignmentCenter;

    [unreadBadge addSubview:unreadCountLabel];
    [unreadCountLabel autoPinHeightToSuperview];
    [unreadCountLabel autoPinWidthToSuperviewWithMargin:3];

    [self addSubview:shadowView];

    [self addSubview:circleView];
    [circleView autoHCenterInSuperview];
    [circleView autoPinEdgeToSuperviewEdge:ALEdgeBottom];

    [shadowView autoPinEdgesToEdgesOfView:circleView];

    [circleView addSubview:iconView];
    [iconView autoCenterInSuperview];

    [self addSubview:unreadBadge];

    [unreadBadge autoPinEdge:ALEdgeBottom toEdge:ALEdgeTop ofView:circleView withOffset:8];
    [unreadBadge autoHCenterInSuperview];
    [unreadBadge autoSetDimension:ALDimensionHeight toSize:16];
    [unreadBadge autoSetDimension:ALDimensionWidth toSize:16 relation:NSLayoutRelationGreaterThanOrEqual];
    [unreadBadge autoMatchDimension:ALDimensionWidth
                        toDimension:ALDimensionWidth
                             ofView:self
                         withOffset:0
                           relation:NSLayoutRelationLessThanOrEqual];
    [unreadBadge autoPinEdgeToSuperviewEdge:ALEdgeTop];

    [self updateColors];
}

- (void)setUnreadCount:(NSUInteger)unreadCount
{
    _unreadCount = unreadCount;

    self.unreadLabel.text = [NSString stringWithFormat:@"%lu", unreadCount];
    self.unreadBadge.hidden = unreadCount < 1;
}

- (void)updateColors
{
    self.unreadBadge.backgroundColor = UIColor.ows_accentBlueColor;
    self.circleView.backgroundColor = Theme.isDarkThemeEnabled ? UIColor.ows_gray65Color : UIColor.ows_gray02Color;
    [self.iconView setTemplateImageName:self.iconName
                              tintColor:Theme.isDarkThemeEnabled ? UIColor.ows_gray15Color : UIColor.ows_gray75Color];
}

@end

NS_ASSUME_NONNULL_END

File: Signal/src/ViewControllers/ConversationView/ConversationScrollButton.swift
//
// Copyright 2022 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

import SignalUI

public class ConversationScrollButton: UIButton {

    let iconName: String

    var unreadCount: UInt = 0 {
        didSet {
            unreadCountLabel.text = String.localizedStringWithFormat("%u", unreadCount)
            unreadBadge.isHidden = unreadCount == 0
        }
    }

    init(iconName: String) {
        self.iconName = iconName
        super.init(frame: .zero)
        createContents()
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(themeDidChange(notification:)),
            name: .ThemeDidChange,
            object: nil)
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    private class var circleSize: CGFloat { ScaleFromIPhone5To7Plus(35, 40) }

    class var buttonSize: CGFloat { circleSize + 2 * 15 }

    private lazy var iconView = UIImageView()

    private lazy var unreadCountLabel: UILabel = {
        let label = UILabel()
        label.font = .systemFont(ofSize: 12)
        label.textColor = .white
        label.textAlignment = .center
        return label
    }()

    private lazy var unreadBadge: UIView = {
        let view = UIView()
        view.isUserInteractionEnabled = false
        view.layer.cornerRadius = 8
        view.clipsToBounds = true
        return view
    }()

    private lazy var shadowView: UIView = {
        let circleSize = ConversationScrollButton.circleSize
        let view = CircleView(diameter: circleSize)
        view.isUserInteractionEnabled = false
        view.layer.shadowOffset = .zero
        view.layer.shadowRadius = 4
        view.layer.shadowOpacity = 0.05
        view.layer.shadowColor = UIColor.black.cgColor
        view.layer.shadowPath = UIBezierPath(ovalIn: CGRect(origin: .zero, size: CGSize(square: circleSize))).cgPath
        return view
    }()

    private lazy var circleView: UIView = {
        let circleSize = ConversationScrollButton.circleSize
        let view = CircleView(diameter: circleSize)
        view.isUserInteractionEnabled = false
        view.layer.shadowOffset = CGSize(width: 0, height: 4)
        view.layer.shadowRadius = 12
        view.layer.shadowOpacity = 0.3
        view.layer.shadowColor = UIColor.black.cgColor
        view.layer.shadowPath = UIBezierPath(ovalIn: CGRect(origin: .zero, size: CGSize(square: circleSize))).cgPath
        return view
    }()

    private func createContents() {
        circleView.addSubview(iconView)
        iconView.autoCenterInSuperview()
        addSubview(shadowView)
        addSubview(circleView)
        circleView.autoHCenterInSuperview()
        circleView.autoPinEdge(toSuperviewEdge: .bottom)
        shadowView.autoPinEdges(toEdgesOf: circleView)

        unreadBadge.addSubview(unreadCountLabel)
        unreadCountLabel.autoPinHeightToSuperview()
        unreadCountLabel.autoPinWidthToSuperview(withMargin: 3)
        addSubview(unreadBadge)
        unreadBadge.autoPinEdge(.bottom, to: .top, of: circleView, withOffset: 8)
        unreadBadge.autoHCenterInSuperview()
        unreadBadge.autoSetDimension(.height, toSize: 16)
        unreadBadge.autoSetDimension(.width, toSize: 16, relation: .greaterThanOrEqual)
        unreadBadge.autoMatch(.width, to: .width, of: self, withOffset: 0, relation: .lessThanOrEqual)
        unreadBadge.autoPinEdge(toSuperviewEdge: .top)

        updateColors()
    }

    private func updateColors() {
        unreadBadge.backgroundColor = .ows_accentBlue
        circleView.backgroundColor = Theme.isDarkThemeEnabled ? .ows_gray65 : .ows_gray02
        iconView.setTemplateImageName(iconName, tintColor: Theme.isDarkThemeEnabled ? .ows_gray15 : .ows_gray75)
    }

    @objc
    private func themeDidChange(notification: Notification) {
        updateColors()
    }
}
